### PR TITLE
Talent Tree Screen Rework

### PR DESCRIPTION
**Pull request by TUsama**

### CODE REVIEW

**Comment by TUsama** [comment]
before:
QQ_1722531723550

after:
QQ_1722531537275

File: src/main/java/com/robertx22/age_of_exile/capability/player/PlayerData.java [hunk 1]
```diff
@@ -172,7 +176,16 @@ public void deserializeNBT(CompoundTag nbt) {
 
         this.team = loadOrBlank(TeamData.class, new TeamData(), nbt, TEAM_DATA, new TeamData());
         this.prophecy = loadOrBlank(PlayerProphecies.class, new PlayerProphecies(), nbt, PROPHECY, new PlayerProphecies());
-        this.talents = loadOrBlank(TalentsData.class, new TalentsData(), nbt, TALENTS_DATA, new TalentsData());
+        TalentsData newTalentsData = loadOrBlank(TalentsData.class, new TalentsData(), nbt, TALENTS_DATA, new TalentsData());
+        var newData = newTalentsData.getPerks().values().stream().map(SchoolData::getAllocatedPoints).collect(Collectors.toSet());
+        var oldData = this.talents.getPerks().values().stream().map(SchoolData::getAllocatedPoints).collect(Collectors.toSet());
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/capability/player/PlayerData.java]: I don't like this. It's here to check if a talent was allocated? Why not just set dirty whenever a talent is added/removed?

File: src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/TalentTree.java [hunk 2]
```diff
@@ -119,6 +123,25 @@ public void onLoadedFromJson() {
 
         grid.loadIntoTree();
 
+        CompletableFuture.runAsync(() -> {
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/TalentTree.java]: Is there a reason this is async? Sorry i'm not familiar with how this works. I can just remove the async part if it breaks?

File: src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/parser/TalentGrid.java [hunk 3]
```diff
@@ -60,21 +61,39 @@ public void loadIntoTree() {
 
         Watch fast2 = new Watch().min(1);
         // this can take a long time if the "hasPath" checks aren't minimized
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/parser/TalentGrid.java]: why are you optimizing this part? This only happens once on world load

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/SkillTreeScreen.java [hunk 4]
```diff
@@ -240,47 +269,6 @@ public void addButtonPublic(AbstractWidget b) {
         this.addRenderableWidget(b);
     }
 
-    private void addConnections() {
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/SkillTreeScreen.java:243]: where did this go?

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/SeparableBufferedImage.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        this.originalImage = originalImage;
+        this.width = originalImage.getWidth();
+        this.height = originalImage.getHeight();
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        ImageIO.write(originalImage, "PNG", byteArrayOutputStream);
+        byte[] bytes = byteArrayOutputStream.toByteArray();
+        byteArrayOutputStream.close();
+        byteArrayOutputStream = null;
+        System.out.println("length is " + bytes.length);
+        this.separateTo = bytes.length / (25 * 1024);
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/SeparableBufferedImage.java]: What are these magic numbers?

**Comment by RobertSkalko** [comment]
Ok i read a bit through it. Honestly is it possible to make this not a code refactor, but an addition?
There's a few pieces of code like the render image parts that I'm pretty sure would have no way of knowing how to port to next minecraft version.

So just not touch any of my current code and add your talent tree in a new place that i'll use as long as it works and i know how to fix it. If it ever breaks, i can then go back to the old tree gui until you can PR a fix?

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/PerkButtonPainter.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        ResourceManager resourceManager = Minecraft.getInstance().getResourceManager();
+
+        try (InputStream colorStream = resourceManager.open(color)) {
+            try (InputStream borderStream = resourceManager.open(border)) {
+                try (InputStream perkStream = resourceManager.open(perk)) {
+                    BufferedImage co = ImageIO.read(colorStream);
+                    BufferedImage bo = ImageIO.read(borderStream);
+                    BufferedImage pe = ImageIO.read(perkStream);
+                    int baseSize = identifier.perk().type.size;
+                    int offColor = (baseSize - 20) / 2;
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/PerkButtonPainter.java]: what's 20

File: src/main/java/com/robertx22/age_of_exile/event_hooks/ontick/OnClientTick.java [hunk 7]
```diff
@@ -30,6 +36,19 @@ public static void onEndTick(Minecraft mc) {
         try {
             Player player = Minecraft.getInstance().player;
 
+            container.values().forEach(x -> {
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/age_of_exile/event_hooks/ontick/OnClientTick.java]: Why would you call talent tree code if the player doesn't even have the tree open?

File: src/main/java/com/robertx22/age_of_exile/capability/player/PlayerData.java (hunk 1 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/capability/player/PlayerData.java]: I think I can do the same trick as the button here, just let the connection renderer wait for the update. 
maybe dirty sync is also a solution, I'll try.

File: src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/TalentTree.java (hunk 2 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/TalentTree.java]: prob not, the handlePaintQueue() on my machine will cost about 1 minute, the game might be completely frozen if this task doesn't run async

File: src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/parser/TalentGrid.java (hunk 3 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/database/data/talent_tree/parser/TalentGrid.java]: yeah I know, but this is also super slow on my PC so I add it on my branch, if it's unnecessary for the public build I can remove it.

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/SkillTreeScreen.java (hunk 4 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/SkillTreeScreen.java:243]: completely replaced by the connection cache.

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/SeparableBufferedImage.java (hunk 5 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/SeparableBufferedImage.java]: well Minecraft can only accept images with byte lengths smaller than 64K, so I have to cut it into several parts

File: src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/PerkButtonPainter.java (hunk 6 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/gui/screens/skill_tree/buttons/drawer/PerkButtonPainter.java]: from the old code.

File: src/main/java/com/robertx22/age_of_exile/event_hooks/ontick/OnClientTick.java (hunk 7 above)
Comment (TUsama) [reply src/main/java/com/robertx22/age_of_exile/event_hooks/ontick/OnClientTick.java]: both tasks can't run if the player hasn't opened the tree screen yet.

**Comment by TUsama** [comment]
> Ok i read a bit through it. Honestly is it possible to make this not a code refactor, but an addition? There's a few pieces of code like the render image parts that I'm pretty sure would have no way of knowing how to port to next minecraft version.
> 
> So just not touch any of my current code and add your talent tree in a new place that i'll use as long as it works and i know how to fix it. If it ever breaks, i can then go back to the old tree gui until you can PR a fix?

I feel that way too, currently the code that triggered these new mechanics is scattered around, which is unhealthy for the project maintenance, I'll find a way to further modulize it.

**Pull request closed without merging**